kong/cluster_events.lua

```diff
@@ -0,0 +1,334 @@
+
+  log(DEBUG, "polling events from: ", min_at, " to: ", max_at)
+
+  for rows, err, page in self.strategy:select_interval(self.channels, min_at, max_at) do
+    if err then
+      return nil, "failed to retrieve events from DB: " .. err
+    end
+
+    if page == 1 then
+      local ok, err = self.shm:safe_set(CURRENT_AT_KEY, max_at)
+      if not ok then
+        return nil, "failed to set 'at' in shm: " .. err
+      end
+    end
+
+    for i = 1, #rows do
+      local row = rows[i]
+
+      if row.node_id ~= self.node_id then
+        local ran, err = self.events_shm:get(row.id)
+        if err then
+          return nil, "failed to probe if event ran: " .. err
+        end
+
+        if not ran then
+          log(DEBUG, "new event (channel: '", row.channel, "') data: '",
+                     row.data, "' nbf: '", row.nbf or "none", "'")
+
+          local exptime = self.poll_interval + self.poll_offset
+
+          -- mark as ran before running in case of long-running callbacks
+          local ok, err = self.events_shm:set(row.id, true, exptime)
+          if not ok then
+            return nil, "failed to mark event as ran: " .. err
+          end
+
+          local cbs = self.callbacks[row.channel]
+          if cbs then
+            for j = 1, #cbs do
+              if not row.nbf then
+                -- unique callback run without delay
+                local ok, err = pcall(cbs[j], row.data)
+                if not ok and not ngx_debug then
+                  log(ERR, "callback threw an error: ", err)
+                end
+
+              else
+                -- unique callback run after some delay
+                local delay = max(row.nbf - ngx_now(), 0)
+
+                log(DEBUG, "delaying nbf event by ", delay, "s")
+
+                local ok, err = timer_at(delay, nbf_cb_handler, cbs[j], row)
+                if not ok then
+                  log(ERR, "failed to schedule nbf event timer: ", err)
+                end
+              end
+            end
+          end
+        end
+      end
+    end
+  end
+
+  return true
+end
+
+
+if ngx_debug then
+  _M.poll = poll
+end
+
+
+local function get_lock(self)
+  local ok, err = self.shm:safe_add(POLL_LOCK_KEY, true, POLL_LOCK_EXPTIME)
+  if not ok and err ~= "exists" then
+    log(ERR, "failed to acquire poll lock: ", err)
+  end
+
+  return ok
+end
+
+
+poll_handler = function(premature, self)
+  if premature or not self.polling then
+    -- set self.polling to false to stop a polling loop
+    return
+  end
+
+  local ok, err = timer_at(self.poll_interval, poll_handler, self)
+  if not ok then
+    log(ERR, "failed to start recurring polling timer: ", err)
+  end
+
+  if not get_lock(self) then
+    return
+  end
+
+  -- single worker
+
+  local pok, perr, err = pcall(poll, self)
+  if not pok then
+    log(ERR, "poll() threw an error: ", perr)
+
+  elseif not perr then
+    log(ERR, "failed to poll: ", err)
+  end
+
+  -- unlock
+
+  self.shm:delete(POLL_LOCK_KEY)
+end
```

Comment:
An alternative approach could be to leverage the `flags` portion with a number representing `ngx.now()` + `poll_interval`, and the lock can only be acquired when the poller's `ngx.now()` is greater than such a value. But this becomes problematic with poll intervals < 1s, and feels kinda fishy.

Alt-alternatively, we can solve the problem of misaligned workers as @Tieske described by forcing each execution to a floor period of time modulo poll interval, e.g., with an interval of 5s, each worker is forced to run at t1, t6, t11, etc... by calculating the timer_at as a diff from `now()` and the next floor up. Of course, both of these scenarios imply that we care about explicitly and consistently running polls exactly at every `poll_interval` seconds- do we?